### DISCUSSION TITLE
Pandoc to PML Coverter Filter

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#26] [General]
I've been considering that, at this stage, it would be useful to provide a way of converting documents in other formats to PML.

The best tool for the task is definitely [pandoc], which supports an [incredible number of input formats].

There are two easy ways to convert from a natively supported pandoc format:

1. [pandoc filters]
1. [pandoc Lua filters]

Both filters work by manipulating an AST intermediate representation of the document (see: [Text-Pandoc-Definition]).

The first one offers complete control over manipulation of the AST, since the filter can be created using any programming language (compiled or scripted). Pandoc passes the AST to the filter as a JSON document.

The second it's easier because pandoc natively supports Lua 5.3 and exposes [a dedicated filter API], which has full access to the document AST; but it also has some limitations.

Unfortunately, I haven't lately been enjoying as much free time as I'd love to, so I didn't manage to dig into more deeply. But I have created a pandoc Lua filter in the past, so I known that it's doable and not that hard either.

I'm not sure whether the Lua filter can adequately cover the need of PML, though, because we'd need to track the various nodes contexts via a stack, in order to properly close tags and handle attributes — maybe it does, but probably it's going to be much simpler to create a custom filter with another language, as long as it has a good JSON parsing library.

Anyhow, this project is still on my wish-list, and when I'll have some decent free time to start working on it I'll give it a try, do some tests and see which approach would be best. For now, I just wanted to share with you my thoughts on this.

Definitely, having a pandoc to PML conversion filter would allow PML users to quickly convert documents form a huge number of formats (Markdown, ODT, etc.) to PML, which is probably a great incentive for those who would like to migrate to PML, since it will spare them all the work of rewriting their docs from scratch.

[pandoc]: https://pandoc.org "pandoc website"
[pandoc filters]: https://pandoc.org/filters.html
[pandoc Lua filters]: https://pandoc.org/lua-filters.html

[incredible number of input formats]: https://pandoc.org/MANUAL.html#option--from

[Text-Pandoc-Definition]: https://hackage.haskell.org/package/pandoc-types-1.22/docs/Text-Pandoc-Definition.html

[a dedicated filter API]: https://pandoc.org/lua-filters.html#module-pandoc

**Comment by celtic-coder** [comment]
Hi Tristano (@tajmone),

Recently, I stumbled across a very interesting Java application called Frink.

It's a full-featured programming language specialising in units conversion. It handles [parsing JSON](https://frinklang.org/#ParsingJSON)  without difficulty.

Kind Regards,
Liam

**Comment by pml-lang** [comment]
> provide a way of converting documents in other formats to PML

Yes, yes, yes!!! I was already thinking about discussing this with you, using Pandoc.

> when I'll have some decent free time to start working on it I'll give it a try

That would be awesome!

> great incentive for those who would like to migrate to PML, since it will spare them all the work of rewriting their docs from scratch.

Agree 1000%.

> Java application called Frink.

@celtic-coder: Thanks for the link.

**Reply by tajmone** [reply]
@celtic-coder:

> Recently, I stumbled across a very interesting Java application called Frink.

Thanks, I'll look into it!

# Errata on Panodc Filters

OK, I've made some confusion about pandoc filters, partly because it was five 'o clock in the morning (no other way to cut free time corners) and partly because I haven't used filters for some years.

A pandoc filter (Lua or otherwise) is intended as an AST processor that intervenes in pandoc's reading (input) or writing (conversion) process of a document format natively supported by pandoc — i.e. you can't use filters to add a new format.

## PML &rarr; HTML &rarr; pandoc

But a pandoc reader filter could be quickly created to aid coversion of PML docs from the PML-generated document to pandoc, and then to any other format supported by pandoc.

Since HTML documents are usually formatted via many arbitrary choices (classes, divs, spans, inline CSS, etc.), it's hard for pandoc to interpret all documents as intended by their author. But since the default PML template is consistent in design, a pandoc Lua input filter could be used to transform the various divs, classes and attributes into pandoc specific AST representation, granting better results in the converted documents.

This is quite easy to do, and I'll soon produce some examples.

## pandoc &rarr; JSON AST &rarr; Custom PML Converter

Pandoc can convert any document to its JSON AST representation, which can then be fed to a custom converter application that parses the JSON file and converts it to an idiomatic PML document — it's just not a "pandoc filter" in this case, I got the terminology wrong in the initial post.

In this case, pandoc Lua filters can also be used to help the conversion process, transforming the AST before feeding it to the converter as a JSON doc — e.g. flatting out or removing unsupported document features (e.g. notes), or carrying out some nodes and attributes conversions to lessen the burden of the custom converter (which would not need to handle unsupported nodes).

So what I said above still applies, just got the terminology mixed up.

# PML Support of Pandoc AST

@pml-lang:

> Yes, yes, yes!!! I was already thinking about discussing this with you, using Pandoc.

If PML is going to support other output formats in the future, levering pandoc is definitely a good choice. PML would only need to translate its internal AST to a pandoc AST in JSON, and this would allow conversion to all the formats natively supported by pandoc (plus the third party converters too).

Also, I guess that since PML already uses its own AST, generating a pandoc JSON AST is going to be less work than implementing native converters to other formats (ODT, markdown, etc.).

Another advantage of this approach is that PML wouldn't have to worry about unsupported features and lossy conversions — pandoc already handles them when converting a JSON AST — so a single JSON output will be used for all formats.

**Reply by pxml-lang** [reply]
This is an important topic. So I want to share what I've learned so far:

# Pandoc features

According to Pandoc's [command line options](https://pandoc.org/MANUAL.html#options):
- the native input format can be 'native Haskell' or 'JSON version of native AST' (`--from / -f` option)
- the native output format can also be 'native Haskell' or 'JSON version of native AST' (`--to / -t` option)

Moreover:
- [filters](https://pandoc.org/filters.html) can be used to change Pandoc's AST (after reading format A, and before writing to format B)
- [custom writers](https://pandoc.org/MANUAL.html#custom-writers) can be written in Lua, and activated on the commend line with the `-t` option. Lua is included in Pandoc (no need for extra installation).

So these are the Pandoc features we can use to convert X to PML, and PML to X.

# Convert X to PML

>  a pandoc Lua input filter could be used to transform the various divs, classes and attributes into pandoc specific AST representation, granting better results in the converted documents.

That's a clever idea! I'm thrilled to see how far we can go with this (probably easiest) solution.

> Pandoc can convert any document to its JSON AST representation, which can then be fed to a custom converter application that parses the JSON file and converts it to an idiomatic PML document

So we have two more options:
- use a [custom writer](https://pandoc.org/MANUAL.html#custom-writers) written in Lua
- create a converter (written in any language) to convert Pandoc's 'native Haskell' or 'JSON version of native AST' to PML.

> In this case, pandoc Lua filters can also be used to help the conversion process

Good idea, too.

# Convert PML to X

> If PML is going to support other output formats in the future, levering pandoc is definitely a good choice.

Agree 100%.

> since PML already uses its own AST, generating a pandoc JSON AST is going to be less work.

Exactly. No need to write a PML to X converter for every X.

> PML wouldn't have to worry about unsupported features and lossy conversions

That's another important advantage.

Although the PML converter currently only writes HTML, it has been designed to add additional writers in the future (to support other output formats).
Hence we just have to add a writer that writes the PML AST to 'native Haskell' or 'JSON version of native AST'. I don't know these formats, so we have to see which one is better suited. And if they are well documented and don't change anymore.

Some interesting links:
[https://pandoc.org/lua-filters.html](https://pandoc.org/lua-filters.html)
[Does Pandoc support custom readers?](https://github.com/jgm/pandoc/issues/3192)
[facility to write custom readers in lua](https://github.com/jgm/pandoc/issues/2311)

**Reply by tajmone** [reply]
> > - custom writers can be written in Lua, and activated on the commend line with the `-t` option. Lua is included in Pandoc (no need for extra installation).

This must be something that was added along the line and I've missed it. Originally, if I remember correctly, filter could only intervene in natively supported formats, but it looks like now things have changed.

I've spent some time sifting through pandoc's Lua API, just to get refreshed on what is being exposed, the various helper functions and variables, and I noticed that the API has been improved a lot, so I guess it should be doable.

Anything that is script based (i.e. not binary compiled) can be tested in the PML Playground repository. So I'll "soon" be setting up prototypes of the various possible solutions.

### Convert X to PML

> That's a clever idea! I'm thrilled to see how far we can go with this (probably easiest) solution.

I agree, we should exploit this path. After all, pandoc will parse an input HTML file into a full pandoc AST, so we only have to ensure that PML classes are transformed into proper AST nodes.

Right now, before starting to experiment with this in the Playground, I've began working on PML sample docs, because we'll need a lot of of them, with increasing complex features (nodes) as the filter grows. For this reason, in the past few days I've began to set-up a dedicated folder for PML sample docs, that will rely on PML "chunks" to be `[include`d in each sample doc (or simply concatenated via shell) so we'll be able to reduce maintenance efforts when PML updates introduce changes to the language — we'll only need to update chunks, and the fixes will propagate to all sample docs used in the repository. The idea behind chunks is that each chunk should provide a sample of a specific feature (lists, tables, admonitions, etc.), so that sample docs can be quickly created by collating various chunks (`[include`/concatenate) according to complexity needs.

This approach might be delaying the initial work, but in the long terms is going to make our life much easier (and DRY).

> Although the PML converter currently only writes HTML, it has been designed to add additional writers in the future (to support other output formats).
Hence we just have to add a writer that writes the PML AST to 'native Haskell' or 'JSON version of native AST'.


I think that pandoc filter are a good starting point, and a way to explore the issue and learn the details. But ultimately, it would be probably best if `pmlc` could natively support generating a pandoc JSON AST and support it as an alternative output to HTML, and possibly also accept a pandoc JSON AST as input (that might be slightly more complex than the former, due to having to disregard unsupported features).

> I don't know these formats, so we have to see which one is better suited.

I think that the JSON AST is the way to go, unless you're actually creating an Haskell app.

> And if they are well documented ...

Documented they are, but in the context of the Haskell API — i.e. there's no language agnostic document listing all nodes in a commented table, indicating which pandoc formats support any given feature, etc.

So I guess will have to just dig through the pandoc types API and start building our own reference docs, as we go along. I'm OK with that, the writing process reinforces the API knowledge, and it would be a great way to pay back the pandoc community to, by sharing the resulting docs.

> ... and don't change anymore.

As for changes, I wouldn't expect big changes really; the pandoc AST (and its types) is already mature enough and supports so many in/out document formats that I believe it already covers the needs of most digital documents in terms of what gets classified as what, and the way it's represented in the AST. Most likely, there could be more elements added (as partially supported formats are further developed), but they wouldn't break backward compatibility, and most likely will reuse some existing nodes types and attributes.

Even the Lua API, I noticed, tends to keep deprecated functions and variables, for the sake of backward compatibility, and simply _discourages_ using them any further. So pandoc development takes into great consideration backward compatibility within a same MAJOR version, and postpones all breaking changes to a next MAJOR release. From that point of view, I would say we can confidently invest energy in pandoc v2.x with expectations that it won't introduce breaking changes we should worry about. Also, each MAJOR incarnations of pandoc lasts for years, so it's not as if every year we'll be witnessing mass migration to a new pandoc version either.

> Some interesting links:

Thanks for the links! Once I manage to set-up the `panodc/` folder in the Playground, I'll add all these links in its README files.

**Reply by celtic-coder** [reply]
Hi Christian (@pml-lang) / Tristano (@tajmone),

I was checking on GitHub for what I could find about the Pandoc AST and came across the [pandoc-lens](https://github.com/bgamari/pandoc-lens) repo. The description says that it is **Lenses for the Pandoc AST**.

I am not familiar with Haskell code, but it looks like [lens.hs](https://github.com/bgamari/pandoc-lens/blob/master/Text/Pandoc/Lens.hs) is breaking down the lenses as  follows:

- [Documents](https://github.com/bgamari/pandoc-lens/blob/07c526cca8aef9b14d2aed3e3d2e080e5104c092/Text/Pandoc/Lens.hs#L11-L14)

![pandoc-documents](https://user-images.githubusercontent.com/11787866/116600497-46192500-a921-11eb-9cd4-c334881b8afa.png)

- [Blocks](https://github.com/bgamari/pandoc-lens/blob/07c526cca8aef9b14d2aed3e3d2e080e5104c092/Text/Pandoc/Lens.hs#L15-L32)

![pandoc-blocks](https://user-images.githubusercontent.com/11787866/116618278-bd0ce880-a936-11eb-9456-01fe0631fe0c.png)

- [Inlines](https://github.com/bgamari/pandoc-lens/blob/master/Text/Pandoc/Lens.hs#L33-L55)

![pandoc-inlines](https://user-images.githubusercontent.com/11787866/116617381-af0a9800-a935-11eb-8bc1-08921d63dde7.png)

- [Metadata](https://github.com/bgamari/pandoc-lens/blob/master/Text/Pandoc/Lens.hs#L56-L65)

![pandoc-metadata](https://user-images.githubusercontent.com/11787866/116617994-67384080-a936-11eb-959d-d10bf3f06308.png)

- [Attributes](https://github.com/bgamari/pandoc-lens/blob/master/Text/Pandoc/Lens.hs#L66-L70)

![pandoc-attributes](https://user-images.githubusercontent.com/11787866/116618731-5c31e000-a937-11eb-94b2-1610d516d77f.png)

The rest of the code is, unfortunately, "gobbledegook" to me. However, I'm wondering if this might be a possible listing of the different document structures that Pandoc understands? I don't know if this would be of any help in working with the AST, but I certainly found the breakdown of the document interesting!

Kind Regards,
Liam

**Reply by celtic-coder** [reply]
It turns out that these are only a small part of the [Text.Pandoc.Definition](https://hackage.haskell.org/package/pandoc-types-1.22/docs/Text-Pandoc-Definition.html). For example, here are the [Block](https://hackage.haskell.org/package/pandoc-types-1.22/docs/Text-Pandoc-Definition.html#t:Block) elements:

![pandoc-block-elements](https://user-images.githubusercontent.com/11787866/116624097-37416b00-a93f-11eb-8d64-6dc6af6551fb.png)

and here are the [Inline](https://hackage.haskell.org/package/pandoc-types-1.22/docs/Text-Pandoc-Definition.html#t:Inline) elements:

![pandoc-inline-elements](https://user-images.githubusercontent.com/11787866/116624343-b2a31c80-a93f-11eb-999e-190f97d495f7.png)

But there are lots more in the definition. I don't know how useful this document might be, but the shorter list above might prove a better starting point, perhaps.

**Reply by pml-lang** [reply]
>  I'll "soon" be setting up prototypes of the various possible solutions.

Great!

> I've began to set-up a dedicated folder for PML sample docs

That's a wonderful idea! I had a quick look at it, and was delighted. Thank you so much, Tristano.
This will also be very useful for users who want to see more examples, or contribute their own ideas. A _one stop place_ for continuously updated PML examples!
Maybe we could write a small tool to automatically add (and then update) all examples from PML's reference manual. These chunks are defined in the converter's source code (here is an [example for `table_data`](https://github.com/pml-lang/converter/blob/main/work/src/PPL/dev/pml/converter/data/formal_node/block/with_raw_text_content/table_data/se_PML_table_data_formal_node.ppl)), so we could programmatically add them, maybe in a folder reserved for the manual's examples, and keep them updated.

> I think that the JSON AST is the way to go, unless you're actually creating an Haskell app.

I agree. Haskell is too 'exotic', and much less popular than JSON.

> it would be a great way to pay back the pandoc community to, by sharing the resulting docs.

Nice and noble thought!

**Reply by pml-lang** [reply]
> It turns out that these are only a small part of the Text.Pandoc.Definition.

Yes, it seems that `Text.Pandoc.Definition` is where we have to look. According to [Using the pandoc API](https://pandoc.org/using-the-pandoc-api.html):
"The Pandoc AST is defined in the pandoc-types package. You should start by looking at the Haddock documentation for Text.Pandoc.Definition."

**Reply by tajmone** [reply]
The pandoc-lens repo (even if it doesn't cover all nodes types) seems a good reference to get an overview on how these different nodes and their attributes are organized (at least, by looking at the snippets pasted by Liam I get a more immediate perception of what the essentials are).

The main problem is understanding which nodes are supported by each format, and how they are actually mapped to the various formats — i.e. some nodes are only supported in some formats (e.g. footnotes are not supported in plain markdown, but they are supported in pandoc markdown flavor, etc. When writing a filter (Lua or otherwise) it might often be necessary to know if a specific node is supported by a given format, because if it isn't the filter might need to convert it to the closest node available, or discard it.

The Lua API exposes to the filter the current target format (input/output), so the hook function for a specific node can handle its contents according to the context.

Our goal when converting from PML to pandoc (via HTML or directly) is to aim for a pure AST representation, but this isn't always possible (e.g. raw HTML code will only be supported by formats which allow HTML, like markdown and other HTML oriented formats) so there should be some strategies on how to handle lossy contents.

**Reply by celtic-coder** [reply]
The [pandoc](https://github.com/jgm/pandoc) repo on GitHub has an extensive [test suite](https://github.com/jgm/pandoc/tree/master/test). This includes a full [HTML test document](https://github.com/jgm/pandoc/blob/master/test/html-reader.html) and the equivalent "[native](https://github.com/jgm/pandoc/blob/master/test/html-reader.native)" AST format. These both handle the "reader" side of things.

Opening the HTML test document in a browser shows what is covered. I don't see a Header Level 6, and it does not appear to test for the modern \<header\>, \<footer\>, \<article\>, \<section\>, \<nav\>, etc. but otherwise it covers much ground. 

Looking at the "native" equivalent would give an idea of what is involved for what would be the bulk of any HTML that PML would output. Having all of this native AST in one document certainly makes it easier to follow the logic of what needs to be written for the equivalent HTML.

There is an [HTML "writer"](https://github.com/jgm/pandoc/blob/master/test/writer.html5) file, but I think that it's purpose is to be a stricter HTML standard as it uses `&quot;` for quotation marks, for example. I guess that the "reader" file does not need to be as strict.

For comparison, there is an [AsciiDoctor writer](https://github.com/jgm/pandoc/blob/master/test/writer.asciidoctor) file to review.

**Reply by celtic-coder** [reply]
Converting from the native AST to an HTML output using `pandoc -f native -t html html-reader.native > native-equivalent.html` produced a slightly different version of the HTML than either the "reader" or "writer" versions. Something to be investigated further at another time.

**Reply by celtic-coder** [reply]
Converting the native AST to JSON (using `pandoc -f native -t json html-reader.native > native-equivalent.json`) gives a single line with over 62k characters.

At least the native AST is readable!  😄

**Reply by pml-lang** [reply]
> The Pandoc repo on GitHub has an extensive test suite ...

> Converting the native AST to JSON ...

Very useful. Thank you.

**Comment by celtic-coder** [comment]
Before I had come across PML., I was starting to do an AsciiDoctor -> DocBook -> Pandoc -> Groff conversion. I have been using LaTeX for quite a number of years and had come across Groff recently and wanted to experiment with it, primarily for PDF output. I know that this looks like a long-winded process, but being able to use Pandoc will be a huge benefit and add greatly to the PML ecosystem. 

Since many already use Pandoc as an intermediate format for the various flavours of Markdown output, it only stands to reason that we should take advantage of this mindset. I therefore share Christian's enthusiasm for this endeavour! 😄 Thanks, Tristano!

**Reply by tajmone** [reply]
> Before I had come across PML., I was starting to do an AsciiDoctor -> DocBook -> Pandoc -> Groff conversion. [...] I know that this looks like a long-winded process, but being able to use Pandoc will be a huge benefit and add greatly to the PML ecosystem.

I'm a great fan of Asciidoctor and use it on daily basis, and unfortunately I know too well that currently the DocBook format is the only viable intermediate format for conversion to formats unsupported by Asciidoctor.

I've been pushing for a native Asciidoctor backend to produce pandoc JSON AST (asciidoctor/asciidoctor#2921) and I believe that we'll be seeing it implemented in the near feature (at the time of my request, Asciidoctor 2.0 was still in the working so it was postponed).

Hopefully, Asciidoctor will also eventually support generating a JSON AST of its own, similarly to pandoc. If all the various documentation tools were to produce JSON AST representation of their documents, it would be much easier to convert to and from these tools my simply manipulating the JSON AST to meet each tool's own criteria.

Who knows ... maybe in the future we'll see the introduction of a new Documentation AST Standard which would allow to represent any document in a format agnostic manner, supporting every possible editorial feature (footnotes, end notes, editor notes, etc.). When this happens, any document conversion tool could then support conversion from any other tool, by simply manipulating the generated AST to adapt it to its own supported features. Although this would mean accepting lossy conversions in most cases, by dropping unsupported features (e.g. cell/row spanning in tables), it would still be more practical than having to parse other markup languages.

Maybe an initial step would be to create a similar standard for lightweight markup syntaxes only, which usually don't cover all the advanced features of digital publishing (e.g. font kerning, advanced page-setting, etc.) found in professional editing tools. Right now, if we had a JSON AST Specification to represent all features supported by Asciidoctor and pandoc Markdown we would already have achieved a good coverage of most markup syntaxes used for collaborative editing of documentation.

By experimenting with pandoc filters and its JSON AST, along with PML and its features, we might eventually gain enough experience to propose an initial draft of such a specification, as a starting point.
